packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -167,6 +215,29 @@ export default createRule<Options, MessageId>({
         if (invalidAncestor.type === AST_NODE_TYPES.ReturnStatement) {
           // handle return statement
 
+          if (options.ignoreVoidInVoid) {
+            const functionNode = findFunction(invalidAncestor);
+
+            if (functionNode != null) {
+              if (
+                (functionNode.returnType &&
+                  functionNode.returnType.typeAnnotation.type !==
+                    AST_NODE_TYPES.TSAnyKeyword &&
+                  functionNode.returnType.typeAnnotation.type !==
+                    AST_NODE_TYPES.TSUnknownKeyword) ||
+                ((functionNode.type === AST_NODE_TYPES.FunctionExpression ||
+                  functionNode.type ===
+                    AST_NODE_TYPES.ArrowFunctionExpression) &&
+                  isValidFunctionExpressionReturnType(functionNode, {
+                    allowTypedFunctionExpressions: true,
+                  })) ||
+                ancestorHasReturnType(functionNode)
```

Comment:
[Refactor] What do you think about moving this condition to the separate function?

That way we can avoid code duplication:

https://github.com/typescript-eslint/typescript-eslint/blob/f392b3961fcb478d4f8f8d2df44301b9dfc0e56a/packages/eslint-plugin/src/rules/no-confusing-void-expression.ts#L148-L158